modules/libraries/src/views/full/UploadLibrary.tsx

```diff
@@ -0,0 +1,111 @@
+import { Button, FileInput, Intent } from '@blueprintjs/core'
+import { lang, toast } from 'botpress/shared'
+import { InstalledLibrary } from 'full'
+import React, { FC } from 'react'
+
+interface Props {
+  lib: InstalledLibrary
+  axios: any
+  refreshLibraries: () => void
+}
+
+const fetchReducer = (state, action) => {
+  if (action.type === 'clearState') {
+    return {
+      file: undefined,
+      fullPath: '',
+      filePath: '',
+      isLoading: false,
+      hasError: false
+    }
+  } else if (action.type === 'uploadCompleted') {
+    return { ...state, isLoading: false }
+  } else if (action.type === 'receivedFile') {
+    const { file, filePath, fullPath } = action.data
+    return { ...state, file, filePath, fullPath }
+  } else if (action.type === 'updateLocation') {
+    const { fullPath, alreadyExists } = action.data
+    return { ...state, fullPath, alreadyExists }
+  } else if (action.type === 'startUpload') {
+    return { ...state, isLoading: true }
+  } else if (action.type === 'showError') {
+    return { ...state, isLoading: false, hasError: true }
+  } else {
+    throw new Error("That action type isn't supported.")
+  }
+}
+
+const UploadLibrary: FC<Props> = props => {
+  const [state, dispatch] = React.useReducer(fetchReducer, {
+    file: undefined,
+    fullPath: '',
+    filePath: '',
+    isLoading: false,
+    hasError: false
+  })
+
+  const { file, fullPath, filePath, isLoading, hasError } = state
+
+  const submitChanges = async () => {
+    dispatch({ type: 'startUpload' })
+
+    try {
+      const form = new FormData()
+      form.append('location', fullPath)
+      form.append('file', file)
+
+      await props.axios.post('/mod/code-editor/upload', form, {
```

Comment:
This is quite wrong IMO.. you created a "circular dependency" between code editor and shared libraries.

Library module uses code editor api .. And code editor can be used to add dependencies ...

I know you won't do anything about this but keep in mind this is a weird practice especially if we want to keep botpress as modular as possible

Reply:
Yup I know, but I don't want to implement all the logic surrounding file upload in this module, because we may miss something in the future. If we move the code editor to the core, then it'll be good.